clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -1249,38 +1255,77 @@ raw2trace_t::open_new_chunk(raw2trace_thread_data_t *tdata)
+            *(int *)buf->encoding);
+        offs += buf->size;
+        size_left -= buf->size;
+        ++buf;
+        CHECK(static_cast<size_t>(buf - buf_start) < WRITE_BUFFER_SIZE,
+              "Too many entries for write buffer");
+    } while (size_left > 0);
+    return "";
+}
+
+std::string
+raw2trace_t::insert_post_chunk_encodings(void *tls, const trace_entry_t *instr,
+                                         app_pc decode_pc)
+{
+    auto tdata = reinterpret_cast<raw2trace_thread_data_t *>(tls);
+    trace_entry_t encodings[WRITE_BUFFER_SIZE];
+    trace_entry_t *buf = encodings;
+    log(4, "Adding post-chunk-boundary encoding entry for decode=%p app=%p\n", decode_pc,
+        instr->addr);
+    std::string err = append_encoding(tls, decode_pc, instr->size, buf, encodings);
+    if (!err.empty())
+        return err;
+    if (!tdata->out_file->write(reinterpret_cast<const char *>(encodings),
+                                reinterpret_cast<const char *>(buf) -
+                                    reinterpret_cast<const char *>(encodings)))
+        return "Failed to write to output file";
+    return "";
+}
+
 // All writes to out_file go through this function, except new chunk headers
 // and footers (to do so would cause recursion; we assume those do not need
 // extra processing here).
 std::string
-raw2trace_t::write(void *tls, const trace_entry_t *start, const trace_entry_t *end)
+raw2trace_t::write(void *tls, const trace_entry_t *start, const trace_entry_t *end,
+                   std::vector<app_pc> decode_pcs)
 {
+    if (end == start)
+        return "Empty buffer passed to write()";
     auto tdata = reinterpret_cast<raw2trace_thread_data_t *>(tls);
     if (tdata->out_archive != nullptr) {
+        bool prev_was_encoding = false;
+        int instr_ordinal = -1;
         for (const trace_entry_t *it = start; it < end; ++it) {
             tdata->cur_chunk_ref_count += tdata->memref_counter.entry_memref_count(it);
-            if (it->type == TRACE_TYPE_MARKER) {
-                if (it->size == TRACE_MARKER_TYPE_TIMESTAMP)
-                    tdata->last_timestamp_ = it->addr;
-                else if (it->size == TRACE_MARKER_TYPE_CPU_ID) {
-                    tdata->last_cpu_ = static_cast<uint>(it->addr);
-                    tdata->sched.emplace_back(tdata->tid, tdata->last_timestamp_,
-                                              tdata->last_cpu_,
-                                              tdata->cur_chunk_instr_count);
-                    tdata->cpu2sched[it->addr].emplace_back(
-                        tdata->tid, tdata->last_timestamp_, tdata->last_cpu_,
-                        tdata->cur_chunk_instr_count);
-                }
-                continue;
-            }
             // We wait until we're past the final instr to write, to ensure we
             // get all its memrefs, by not stopping until we hit an instr or an
             // encoding.  (We will put function markers for entry in the
```

Comment:
Unrelated to this PR: so we duplicate function markers? I remember a discussion on that recently. Is this the cause?

Reply:
No, we don't duplicate function markers.  Not sure which discussion: if you remember more we can discuss ofline.